### ISSUE TITLE
Enabling firewall blocks Cockpit

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit#17660]
### Explain what happens

1. Visit the firewall page
2. Enable the firewall
3. Lose access to Cockpit

This is apparently a problem on at least:

- openSUSE Tumbleweed
- Fedora IoT

And it might be a problem elsewhere too. It's worth looking to see if it affects RHEL, CentOS, and Fedora Server.

Cockpit is supposed to be allowed by default. This is either a regression in Cockpit or an assumption made in the default packages and configuration.

If the default zone doesn't allow for it, then we need to enable Cockpit before turning on the firewall. Perhaps we need a prompt about Cockpit not being in the zone, so it will be added so that a connection to Cockpit will not be lost.

This would only affect Cockpit when connecting through the socket, not via SSH (like with Cockpit Client or via another host), so it should take this into account. In other words, if someone is connecting through SSH (for example), then turning on the firewall won't stop the connection to Cockpit (unless ssh isn't allowed through the firewall) and therefore shouldn't show a confirmation prompt that adds Cockpit.

### Version of Cockpit

main, 273

### Server operating system

openSUSE

**Comment by garrett** [comment]
@skazi0 brought this to our attention in IRC today, specifically about openSUSE, also pointing out that our tests have a workaround, indicating that this is a problem at least on some of the distributions: https://github.com/cockpit-project/cockpit/blob/main/test/vm.install#L7

So while we do a preventive check for this when removing Cockpit, we don't check for turning on the firewall without Cockpit enabled. And you cannot add Cockpit until you turn the firewall on, hence my suggestion about Cockpit adding itself to the zones between enabling and turning it on, so the user isn't blocked.

**Comment by MahmoudHamdy02** [comment]
Automatically adding the cockpit service to the public zone seems to work OK, this line adds the cockpit service every time the firewall is turned on.

https://github.com/cockpit-project/cockpit/blob/main/pkg/networkmanager/firewall-client.js#L353
`firewall.enable = () => firewalld_service.enable().then(() => firewalld_service.start()).then(() => { firewall.addService("public", "cockpit") });`

Let me know your thoughts about this fix!

**Comment by MahmoudHamdy02** [comment]
I'm on the third point now, but had a quick question.
Is there a way to access the active services when the firewall is disabled?

Otherwise it would maybe have to go like this:
Enable firewall -> check if cockpit is active -> if not, disable firewall and show a modal -> add cockpit and enable firewall again

**Comment by martinpitt** [comment]
> Is there a way to access the active services when the firewall is disabled?

No, I don't think so. cockpit needs the D-Bus interface of firewalld, so does `firewall-cmd`.  I don't know if firewalld only blocks *new* TCP connections (TCP SYN) or established ones as well -- i.e. I am not sure if a running cockpit session would actually get broken by enabling firewalld while it's running. Easy enough to test, please let us know here.

Your approach seems okay. Depending on the above, the "check if cockpit is active"  could either be "opening a new channel works" (if the existing websocket connection works) or "pinging the cockpit socket works", or simply "ask firewalld about the cockpit service" (but that would be wrong if the admin changed the port to something other than 9090, or is connecting via SSH). This is still a bit hand-wavy, but let's do the above test first.

**Comment by MahmoudHamdy02** [comment]
>  I am not sure if a running cockpit session would actually get broken by enabling firewalld while it's running. 

I have tested it and this seems to be the case, when I enable the firewall from my phone, the connection is still enabled. (For example I can add/remove a service from a zone). Once I switch to another page the connection breaks.

I will keep looking into this and let you know if I reach a working solution.

**Comment by MahmoudHamdy02** [comment]
Since the connection doesn't die instantly, I guess the easiest solution (or at least a starting point) is to check if cockpit is in the services list when firewall is turned on. If it doesn't exist, show a message like "The default cockpit port is not allowed in the firewall, which might cause connection loss. Would you like to add it now?" with Add and "x"/dismiss buttons
